packages/zengine-sdk/src/filter-matcher.ts

```diff
@@ -0,0 +1,263 @@
+        recordValue = parseFloat(recordValue)
+      } catch (err) {
+        return false
+      }
+
+      return recordValue >= ruleValue
+    }
+    // Loose comparison, don't use 'localeCompare' because we only want to compare
+    // numbers or well-formatted dates, which work with normal string comparison.
+    return recordValue >= ruleValue
+  },
+  ruleMaximum: function (recordValue: any, ruleValue: any) {
+    if (Array.isArray(recordValue)) {
+      return false
+    }
+    if (recordValue === '') {
+      return false
+    }
+    // Do numeric comparison if filter rule value is numeric. For consistency with API,
+    // record values are somewhat forgiving, while filter rules must be a valid number.
+    if (ruleValue && !isNaN(Number(ruleValue))) {
+      recordValue = parseNumber(recordValue)
+
+      try {
+        ruleValue = parseFloat(ruleValue)
+        recordValue = parseFloat(recordValue)
+      } catch (err) {
+        return false
+      }
+
+      return recordValue <= ruleValue
+
+    }
+    // Loose comparison, don't use 'localeCompare' because we only want to compare
+    // numbers or well-formatted dates, which work with normal string comparison.
+    return recordValue <= ruleValue
+  },
+  ruleContains: function (recordValue: any, ruleValue: any) {
+    if (Array.isArray(recordValue)) {
+
+      // Normalize Value as Array
+      if (!Array.isArray(ruleValue)) {
+        ruleValue = [ruleValue]
+      }
+
+      for (let index = 0; index < ruleValue.length; index++) {
+        if (recordValue.indexOf(ruleValue[index]) === -1) {
+          return false
+        }
+      }
+      return true
+    } else {
+      return recordValue.toString().indexOf(ruleValue.toString()) !== -1
+    }
+  },
+  ruleDoesNotContain: function (recordValue: any, ruleValue: any) {
+    return !matchers.ruleContains(recordValue, ruleValue)
+  },
+  ruleStartsWith: function (recordValue: any, ruleValue: any) {
+    if (Array.isArray(recordValue)) {
+      return false
+    }
+    return recordValue.toString().startsWith(ruleValue.toString())
+  },
+  ruleEndsWith: function (recordValue: any, ruleValue: any) {
+    if (Array.isArray(recordValue)) {
+      return false
+    }
+    return recordValue.toString().endsWith(ruleValue.toString())
+  },
+  ruleIn: function (recordValue: any, ruleValue: any) {
+    if (!Array.isArray(ruleValue)) {
+      return false
+    }
+    if (Array.isArray(recordValue)) {
+      for (let i = 0; i < recordValue.length; i++) {
+        if (matchers.ruleIn(recordValue[i], ruleValue)) {
+          return true
+        }
+      }
+      return false
+    } else {
+      const iRecordValue = recordValue.toString().toLowerCase()
+      const iRuleValues = ruleValue.map(function (value) {
+        return value.toString().toLowerCase()
+      })
+      return iRuleValues.indexOf(iRecordValue) !== -1
+    }
+  },
+  ruleNotIn: function (recordValue: any, ruleValue: any) {
+    if (!Array.isArray(ruleValue)) {
```

Comment:
This check is duplicated, since this is the opening line of `ruleIn`, too.

Reply:
correct, this was copied as-is from anglerfish.  since we're negating the value of `ruleIn` to determine `ruleNotIn` if the string was empty we'd be returning true here though